### PR TITLE
docs(tutorial): add new tutorial for serving custom models in Instill Core

### PR DESCRIPTION
**Pull request by GeorgeWilliamStrong**
Because

- We need a new tutorial on creating custom models on Instill Model 

This commit

- Creates the new tutorial

### CODE REVIEW

**Review by orangecoloured — APPROVED** [review]
Code-wise LGTM.

**Review by heiruwu — DISMISSED** [review]
The tutorial looks awesome!

File: tutorials/model-serving-on-instill-core.mdx [hunk 1]
```diff
@@ -0,0 +1,544 @@
...
+Core**.
+
+## Prerequisites
+
+1. **Docker**: Both **🔮 Instill Core** and **⚗️ Instill Model** use **Docker**
+   to ensure that models and code can be deployed in consistent, isolated and
+   reproducible environments. Please ensure that you have Docker installed and
+   running by following the [official
+   instructions](https://docs.docker.com/get-docker/), and see our [deployment
+   guide](./docs/core/deployment) for recommended resource settings.
```
Comment (pinglin) [inline tutorials/model-serving-on-instill-core.mdx:51]:
```suggestion
   guide](/docs/core/deployment) for recommended resource settings.
```

File: tutorials/model-serving-on-instill-core.mdx [hunk 2]
```diff
@@ -0,0 +1,544 @@
...
+4. Now that **🔮 Instill Core** has been deployed, you can access **📺 Instill
+   Console** at [http://localhost:3000](http://localhost:3000). Please use the
+   following initial login details to initiate the password reset process for
+   onboarding:
+
+    - Username: `admin`
+    - Password: `password`
+
+For further details about launching **🔮 Instill Core**, we recommend that you
+refer to the [deployment guide](./docs/core/deployment) which covers alternative
```
Comment (pinglin) [inline tutorials/model-serving-on-instill-core.mdx:74]:
```suggestion
refer to the [deployment guide](/docs/core/deployment) which covers alternative
```

File: tutorials/model-serving-on-instill-core.mdx [hunk 3]
```diff
@@ -0,0 +1,544 @@
...
+fields, please refer to the [Create a Model](/docs/model/create) page.
+
+In this tutorial, we will be walking through how to create and deploy a version
+of the
+[TinyLlama-1.1B-Chat](https://huggingface.co/TinyLlama/TinyLlama-1.1B-Chat-v1.0)
+model. To follow along, please fill in the configuration fields as per the image
+below.
+
+<ZoomableImg src="/tutorial-assets/model-serving-on-instill-core/T1.png" alt="2.
+Configure Model Settings" />
```
Comment (pinglin) [inline tutorials/model-serving-on-instill-core.mdx:102]:
```suggestion
<ZoomableImg src="/tutorial-assets/model-serving-on-instill-core/T1.png" alt="Configure Model Settings" />
```

File: tutorials/model-serving-on-instill-core.mdx [hunk 4]
```diff
@@ -0,0 +1,544 @@
...
+<CH.Code>
+```yaml instill.yaml
+build:
+  gpu: true
+  python_version: "3.11"
+  python_packages:
+     - torch==2.2.1
+     - transformers==4.36.2
+     - accelerate==0.25.0
+repo: {user_id}/{model_id}
```
Comment (pinglin) [inline tutorials/model-serving-on-instill-core.mdx:127]:
```suggestion
repo: USER_ID/MODEL_ID
```

File: tutorials/model-serving-on-instill-core.mdx [hunk 5]
```diff
@@ -0,0 +1,544 @@
...
+```
+</CH.Code>
+
+Importantly, you must replace `{user_id}` with `admin`, and replace `{model_id}`
```
Comment (pinglin) [inline tutorials/model-serving-on-instill-core.mdx:131]:
```suggestion
Importantly, you must replace `USER_ID` with `admin`, and replace `MODEL_ID`
```

, to align with the [doc](https://www.instill.tech/docs/model/prepare#model-configuration).

File: tutorials/model-serving-on-instill-core.mdx [hunk 6]
```diff
@@ -0,0 +1,544 @@
...
+In the second phase, we define the `ModelMetadata` method which is responsible
+for communicating the models expected input and output shapes to the backend
+service. To easily facilitate this, we can make use of the [Python
+SDK](/docs/latest/sdk/python-sdk) through the `instill.helpers` module which
+provides a number of functions that can be selected according to the AI task the
+model performs.
+
+Here, we recognise that the
+[TinyLlama-1.1B-Chat](https://huggingface.co/TinyLlama/TinyLlama-1.1B-Chat-v1.0)
+model falls under the `text_generation_chat` AI task, and so we will make use of
```
Comment (pinglin) [inline tutorials/model-serving-on-instill-core.mdx:191]:
```suggestion
model falls under the [Text Generation Chat](https://www.instill.tech/docs/model/ai-task#text-generation-chat) AI task, and so we will make use of
```
, to cite the AI task reference.

File: tutorials/model-serving-on-instill-core.mdx [hunk 7]
```diff
@@ -0,0 +1,544 @@
...
+### Implement the Inference Method
+
+In the third phase, we implement the inference method `__call__`, which handles
+the trigger request from **⚗️ Instill Model**, contains the necessary logic to
+run the inference, and constructs the response. We use the `StandardTaskIO`
+module to parse the request payload into input parameters, and convert the model
+outputs to the appropriate response format.
+
+The `TextGenerationChatInput` class from `instill.helpers.const` is used to
+define the input format for the text generation chat AI task, and the
```
Comment (pinglin) [inline tutorials/model-serving-on-instill-core.mdx:222]:
```suggestion
define the input format for the [Text Generation Chat](https://www.instill.tech/docs/model/ai-task#text-generation-chat) AI task, and the
```

File: tutorials/model-serving-on-instill-core.mdx [hunk 8]
```diff
@@ -0,0 +1,544 @@
...
+            do_sample=True,
+            temperature=task_text_generation_chat_input.temperature,
+            top_k=task_text_generation_chat_input.top_k,
+            top_p=0.95,
+        )
+
+        # convert the model output into response output
+        task_text_generation_chat_output = (
+            StandardTaskIO.parse_task_text_generation_chat_output(sequences=sequences)
+        )
```
Comment (pinglin) [inline tutorials/model-serving-on-instill-core.mdx:278]:
```suggestion
        task_text_generation_chat_output: TextGenerationChatInput = (
            StandardTaskIO.parse_task_text_generation_chat_output(sequences=sequences)
        )
```

I supposed the type hint should work as in the input case?

File: tutorials/model-serving-on-instill-core.mdx [hunk 9]
```diff
@@ -0,0 +1,544 @@
...
```
Comment (pinglin) [inline tutorials/model-serving-on-instill-core.mdx:360]:
```suggestion
        task_text_generation_chat_output: TextGenerationChatInput = (
            StandardTaskIO.parse_task_text_generation_chat_output(sequences=sequences)
        )
```

File: tutorials/model-serving-on-instill-core.mdx [hunk 10]
```diff
@@ -0,0 +1,544 @@
...
+completion, you should see a similar output to the following:
+```
+2024-05-28 01:54:44,404.404 INFO     [Instill Builder] admin/tinyllama:v1 built
+2024-05-28 01:54:44,423.423 INFO     [Instill Builder] Done
+```
+
+### Push the Model Image
+
+To push the model image to the **🔮 Instill Core** instance we will need to be
+able to login to Docker. To do this, you first need to create an API token by:
```
Comment (pinglin) [inline tutorials/model-serving-on-instill-core.mdx:411]:
```suggestion
able to login to the hosted Docker registry. To do this, you first need to create an API token by:
```

File: tutorials/model-serving-on-instill-core.mdx [hunk 11]
```diff
@@ -0,0 +1,544 @@
...
+1. Selecting the profile icon in the top right corner of the Console window and
+   choosing the Settings option.
+2. Select API Tokens from the left-hand menu.
+3. Click the **Create Token** button and give it a name, e.g. `tutorial`. Copy
+   the generated API token.
+
+<ZoomableImg src="/tutorial-assets/model-serving-on-instill-core/T2.png" alt="2.
+Create API Token" />
```
Comment (pinglin) [inline tutorials/model-serving-on-instill-core.mdx:419]:
```suggestion
<ZoomableImg src="/tutorial-assets/model-serving-on-instill-core/T2.png" alt="Create API Token" />
```

File: tutorials/model-serving-on-instill-core.mdx [hunk 12]
```diff
@@ -0,0 +1,544 @@
...
+
+Now we can login to Docker on the **🔮 Instill Core** instance by running:
```
Comment (pinglin) [inline tutorials/model-serving-on-instill-core.mdx:421]:
```suggestion
Now we can login to the Docker register in the **🔮 Instill Core** instance by running:
```

File: tutorials/model-serving-on-instill-core.mdx [hunk 13]
```diff
@@ -0,0 +1,544 @@
...
+You should see a response generated by the
+[TinyLlama-1.1B-Chat](https://huggingface.co/TinyLlama/TinyLlama-1.1B-Chat-v1.0)
+model in the **Output** pane.
+
+To access model inferences via the API:
+1. Navigate to the **API** tab for your `admin/tinyllama` model.
+2. Follow the instructions by setting your own `INSTILL_API_TOKEN` as an
+   environment variable:
+   ```bash
+   $ export INSTILL_API_TOKEN=********
```
Comment (pinglin) [inline tutorials/model-serving-on-instill-core.mdx:482]:
```suggestion
   export INSTILL_API_TOKEN=********
```

, to align with the other command line code snippets.

**Review by pinglin — CHANGES_REQUESTED** [review]
The content is fantastic and well-written. 

Besides some minor layout and fixing links requests further, can we also add a section `Step 6. Tear Everything Down` to guide the users how to undeploy a model and tear down the Instill Core instance by running:
```
inst local undeploy
```

File: tutorials/model-serving-on-instill-core.mdx (hunk 8 above)
Comment (GeorgeWilliamStrong) [reply tutorials/model-serving-on-instill-core.mdx:278]: I don't think so, I think for the output case this returns `serialized_bytes_tensor`.

File: tutorials/model-serving-on-instill-core.mdx (hunk 9 above)
Comment (GeorgeWilliamStrong) [reply tutorials/model-serving-on-instill-core.mdx:360]: I don't think so, I think for the output case this returns `serialized_bytes_tensor`.

Comment (pinglin) [reply tutorials/model-serving-on-instill-core.mdx:359]: Hey @heiruwu, can you help confirm this? There is a type `TextGenerationChatOutput` in the [protogen-python](https://github.com/instill-ai/protogen-python/blob/main/model/model/v1alpha/task_text_generation_chat_pb2.pyi#L89), and I feel the input and output expression here could be symmetric.

Comment (heiruwu) [reply tutorials/model-serving-on-instill-core.mdx:364]: no, the `task_text_generation_chat_output` here is the serialized bye string that will be passed back to model-backend, not the user.
the `TextGenerationChatOutput` class is the response for the trigger endpoint.

Comment (heiruwu) [reply tutorials/model-serving-on-instill-core.mdx:364]: the `TextGenerationChatInput` class is written [here](https://github.com/instill-ai/python-sdk/blob/main/instill/helpers/const.py#L58-L68) for the user's convenience to understand the request data structure after its been deserialized, not by the protogen-go

Comment (GeorgeWilliamStrong) [reply tutorials/model-serving-on-instill-core.mdx:364]: @pinglin @heiruwu are we ok to resolve and merge?

**Pull request merged**